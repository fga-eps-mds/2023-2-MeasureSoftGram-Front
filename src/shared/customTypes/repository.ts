export interface TsqmiHistory {
  id: number;
  value: number;
  created_at: Date;
}

export interface Historical {
  id: number;
  key: string;
  name: string;
  history: Array<TsqmiHistory>;
  latest: Result;
  goal?: number;
}

export interface HistoricalCharacteristicsProps {
  organizationId: string | undefined;
  productId: string | undefined;
  repositoryId: string | undefined;
  entity: string;
}

export interface Result {
  id: number;
  value: number;
  created_at: Date;
}

export interface RootObject {
  count: number;
  next?: any;
  previous?: any;
  results: Result[];
}

interface RepositoryResults {
  metrics: string;
  measures: string;
  subcharacteristics: string;
  characteristics: string;
  tsqmi: string;
}

export interface Repository {
  id: number;
  url: string;
  name: string;
  key: string;
  description: string;
  product: string;
  latest_values: RepositoryResults;
  historical_values: RepositoryResults;
}
