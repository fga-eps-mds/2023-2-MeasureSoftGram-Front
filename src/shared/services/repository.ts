/* eslint-disable class-methods-use-this */
import * as sqcHistorical from '@customTypes/sqcHistorical';
import api from './api';

interface HistoricalCharacteristicsProps {
  organizationId: number;
  productId: number | undefined;
  repositoryId: number | undefined;
  entity: string;
}

class Repository {
  getRepository(organizationId: string, productId: number, repositoryId: number) {
    return api.get(`organizations/${organizationId}/products/${productId}/repositories/${repositoryId}/`);
  }

  getHistoricalCharacteristics(props: HistoricalCharacteristicsProps) {
    const { organizationId, entity, productId, repositoryId } = props;
    return api.get(
      `organizations/${organizationId}` +
        `/products/${productId}/repositories/${repositoryId}` +
        `/historical-values/${entity}/`
    );
  }

  getSqcHistory(organizationId: string, productId: string | number, repositoryId: number | undefined) {
    const url = `organizations/${organizationId}/products/${productId}/repositories/${repositoryId}/historical-values/sqc/`;
    return api.get<sqcHistorical.RootObject>(url);
  }
}

export const repository = new Repository();
Object.freeze(repository);
