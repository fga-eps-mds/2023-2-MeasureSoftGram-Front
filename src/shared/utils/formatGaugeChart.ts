import { Historical } from '@customTypes/repository';

export interface FormatGaugeChartType {
  historical?: Historical[];
  title: string;
  isEmpty: boolean;
}

const formatGaugeChart = ({ historical, title, isEmpty }: FormatGaugeChartType) => {
  const vertical: string = `57%`;
  const historicalLength: number = historical?.length ?? 1;

  let incrementX: number = 25;
  if (historicalLength % 2 === 0) {
    incrementX = 50;
  }

  const seriesData =
    historical?.map((item, index) => {
      const horizontal: string = `${25 + index * incrementX}%`;

      return {
        name: item.name,
        show: !isEmpty,
        center: [horizontal, vertical],
        min: 0,
        max: 1,
        data: [
          {
            value: item.latest.value.toFixed(2),
            name: item.name,
            title: {
              offsetCenter: ['-60%', '70%']
            },
            detail: {
              offsetCenter: ['-60%', '95%']
            }
          },
          {
            value: ((item?.goal ?? 1) / 100)?.toFixed(2),
            name: 'Valor esperado',
            title: {
              offsetCenter: ['60%', '70%']
            },
            detail: {
              offsetCenter: ['60%', '95%']
            },
            itemStyle: {
              color: '#f1c40f'
            }
          }
        ],
        type: 'gauge',
        startAngle: 210,
        endAngle: -30,
        anchor: {
          show: true,
          showAbove: true,
          size: 18,
          itemStyle: {
            color: '#FAC858'
          }
        },
        pointer: {
          icon: 'path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z',
          width: 8,
          length: '80%',
          offsetCenter: [0, '8%']
        },
        progress: {
          show: true,
          overlap: true,
          roundCap: true
        },
        axisLine: {
          roundCap: true
        },
        detail: {
          width: 20,
          height: 14,
          fontSize: 18,
          color: '#fff',
          backgroundColor: 'inherit',
          borderRadius: 3
        }
      };
    }) ?? [];

  return {
    title: {
      text: title
    },
    series: seriesData
  };
};

export default formatGaugeChart;
