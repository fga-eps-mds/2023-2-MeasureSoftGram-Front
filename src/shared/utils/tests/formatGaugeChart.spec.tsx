import { Historical } from '@customTypes/repository';
import formatGaugeChart, { FormatGaugeChartType } from '@utils/formatGaugeChart';

describe('formatGaugeChart', () => {

  const chartTitle:string = 'title-test';
  it('should return an object with the correct format', () => {
    const historic: Historical = {
      id: 1,
      key: 'reliability',
      name: 'Reliability',
      history: [],
      latest: {
        id: 12,
        value: 32,
        created_at: new Date(),
      }
    }
    const historical: Historical[] = [
      historic,
      historic,
      historic,
    ];

    const params: FormatGaugeChartType = {
      historical,
      title: chartTitle,
      isEmpty: false
    }

    const values = formatGaugeChart(params);

    expect(values).toMatchSnapshot();
  });

  it('should return an object with the correct format when historical length is even', () => {
    const historic: Historical = {
      id: 1,
      key: 'reliability',
      name: 'Reliability',
      history: [],
      latest: {
        id: 12,
        value: 32,
        created_at: new Date(),
      }
    }
    const historical: Historical[] = [
      historic,
      historic,
    ];

    const params: FormatGaugeChartType = {
      historical,
      title: chartTitle,
      isEmpty: false
    }

    const values = formatGaugeChart(params);

    expect(values).toMatchSnapshot();
  });

  it('should return an object considering empty historic', () => {

    const params: FormatGaugeChartType = {
      historical: undefined,
      title: 'chartTitle',
      isEmpty: false
    }

    const values = formatGaugeChart(params);

    expect(values).toMatchSnapshot();
  });
});
