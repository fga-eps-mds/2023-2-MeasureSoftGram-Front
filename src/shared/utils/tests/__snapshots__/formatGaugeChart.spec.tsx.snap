// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatGaugeChart should return an object considering empty historic 1`] = `
Object {
  "series": Array [],
  "title": Object {
    "text": "chartTitle",
  },
}
`;

exports[`formatGaugeChart should return an object with the correct format 1`] = `
Object {
  "series": Array [
    Object {
      "anchor": Object {
        "itemStyle": Object {
          "color": "#FAC858",
        },
        "show": true,
        "showAbove": true,
        "size": 18,
      },
      "axisLine": Object {
        "roundCap": true,
      },
      "center": Array [
        "25%",
        "57%",
      ],
      "data": Array [
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "-60%",
              "95%",
            ],
          },
          "name": "Reliability",
          "title": Object {
            "offsetCenter": Array [
              "-60%",
              "70%",
            ],
          },
          "value": "32.00",
        },
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "60%",
              "95%",
            ],
          },
          "itemStyle": Object {
            "color": "#f1c40f",
          },
          "name": "Valor esperado",
          "title": Object {
            "offsetCenter": Array [
              "60%",
              "70%",
            ],
          },
          "value": "0.01",
        },
      ],
      "detail": Object {
        "backgroundColor": "inherit",
        "borderRadius": 3,
        "color": "#fff",
        "fontSize": 18,
        "height": 14,
        "width": 20,
      },
      "endAngle": -30,
      "max": 1,
      "min": 0,
      "name": "Reliability",
      "pointer": Object {
        "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
        "length": "80%",
        "offsetCenter": Array [
          0,
          "8%",
        ],
        "width": 8,
      },
      "progress": Object {
        "overlap": true,
        "roundCap": true,
        "show": true,
      },
      "show": true,
      "startAngle": 210,
      "type": "gauge",
    },
    Object {
      "anchor": Object {
        "itemStyle": Object {
          "color": "#FAC858",
        },
        "show": true,
        "showAbove": true,
        "size": 18,
      },
      "axisLine": Object {
        "roundCap": true,
      },
      "center": Array [
        "50%",
        "57%",
      ],
      "data": Array [
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "-60%",
              "95%",
            ],
          },
          "name": "Reliability",
          "title": Object {
            "offsetCenter": Array [
              "-60%",
              "70%",
            ],
          },
          "value": "32.00",
        },
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "60%",
              "95%",
            ],
          },
          "itemStyle": Object {
            "color": "#f1c40f",
          },
          "name": "Valor esperado",
          "title": Object {
            "offsetCenter": Array [
              "60%",
              "70%",
            ],
          },
          "value": "0.01",
        },
      ],
      "detail": Object {
        "backgroundColor": "inherit",
        "borderRadius": 3,
        "color": "#fff",
        "fontSize": 18,
        "height": 14,
        "width": 20,
      },
      "endAngle": -30,
      "max": 1,
      "min": 0,
      "name": "Reliability",
      "pointer": Object {
        "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
        "length": "80%",
        "offsetCenter": Array [
          0,
          "8%",
        ],
        "width": 8,
      },
      "progress": Object {
        "overlap": true,
        "roundCap": true,
        "show": true,
      },
      "show": true,
      "startAngle": 210,
      "type": "gauge",
    },
    Object {
      "anchor": Object {
        "itemStyle": Object {
          "color": "#FAC858",
        },
        "show": true,
        "showAbove": true,
        "size": 18,
      },
      "axisLine": Object {
        "roundCap": true,
      },
      "center": Array [
        "75%",
        "57%",
      ],
      "data": Array [
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "-60%",
              "95%",
            ],
          },
          "name": "Reliability",
          "title": Object {
            "offsetCenter": Array [
              "-60%",
              "70%",
            ],
          },
          "value": "32.00",
        },
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "60%",
              "95%",
            ],
          },
          "itemStyle": Object {
            "color": "#f1c40f",
          },
          "name": "Valor esperado",
          "title": Object {
            "offsetCenter": Array [
              "60%",
              "70%",
            ],
          },
          "value": "0.01",
        },
      ],
      "detail": Object {
        "backgroundColor": "inherit",
        "borderRadius": 3,
        "color": "#fff",
        "fontSize": 18,
        "height": 14,
        "width": 20,
      },
      "endAngle": -30,
      "max": 1,
      "min": 0,
      "name": "Reliability",
      "pointer": Object {
        "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
        "length": "80%",
        "offsetCenter": Array [
          0,
          "8%",
        ],
        "width": 8,
      },
      "progress": Object {
        "overlap": true,
        "roundCap": true,
        "show": true,
      },
      "show": true,
      "startAngle": 210,
      "type": "gauge",
    },
  ],
  "title": Object {
    "text": "title-test",
  },
}
`;

exports[`formatGaugeChart should return an object with the correct format when historical length is even 1`] = `
Object {
  "series": Array [
    Object {
      "anchor": Object {
        "itemStyle": Object {
          "color": "#FAC858",
        },
        "show": true,
        "showAbove": true,
        "size": 18,
      },
      "axisLine": Object {
        "roundCap": true,
      },
      "center": Array [
        "25%",
        "57%",
      ],
      "data": Array [
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "-60%",
              "95%",
            ],
          },
          "name": "Reliability",
          "title": Object {
            "offsetCenter": Array [
              "-60%",
              "70%",
            ],
          },
          "value": "32.00",
        },
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "60%",
              "95%",
            ],
          },
          "itemStyle": Object {
            "color": "#f1c40f",
          },
          "name": "Valor esperado",
          "title": Object {
            "offsetCenter": Array [
              "60%",
              "70%",
            ],
          },
          "value": "0.01",
        },
      ],
      "detail": Object {
        "backgroundColor": "inherit",
        "borderRadius": 3,
        "color": "#fff",
        "fontSize": 18,
        "height": 14,
        "width": 20,
      },
      "endAngle": -30,
      "max": 1,
      "min": 0,
      "name": "Reliability",
      "pointer": Object {
        "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
        "length": "80%",
        "offsetCenter": Array [
          0,
          "8%",
        ],
        "width": 8,
      },
      "progress": Object {
        "overlap": true,
        "roundCap": true,
        "show": true,
      },
      "show": true,
      "startAngle": 210,
      "type": "gauge",
    },
    Object {
      "anchor": Object {
        "itemStyle": Object {
          "color": "#FAC858",
        },
        "show": true,
        "showAbove": true,
        "size": 18,
      },
      "axisLine": Object {
        "roundCap": true,
      },
      "center": Array [
        "75%",
        "57%",
      ],
      "data": Array [
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "-60%",
              "95%",
            ],
          },
          "name": "Reliability",
          "title": Object {
            "offsetCenter": Array [
              "-60%",
              "70%",
            ],
          },
          "value": "32.00",
        },
        Object {
          "detail": Object {
            "offsetCenter": Array [
              "60%",
              "95%",
            ],
          },
          "itemStyle": Object {
            "color": "#f1c40f",
          },
          "name": "Valor esperado",
          "title": Object {
            "offsetCenter": Array [
              "60%",
              "70%",
            ],
          },
          "value": "0.01",
        },
      ],
      "detail": Object {
        "backgroundColor": "inherit",
        "borderRadius": 3,
        "color": "#fff",
        "fontSize": 18,
        "height": 14,
        "width": 20,
      },
      "endAngle": -30,
      "max": 1,
      "min": 0,
      "name": "Reliability",
      "pointer": Object {
        "icon": "path://M2.9,0.7L2.9,0.7c1.4,0,2.6,1.2,2.6,2.6v115c0,1.4-1.2,2.6-2.6,2.6l0,0c-1.4,0-2.6-1.2-2.6-2.6V3.3C0.3,1.9,1.4,0.7,2.9,0.7z",
        "length": "80%",
        "offsetCenter": Array [
          0,
          "8%",
        ],
        "width": 8,
      },
      "progress": Object {
        "overlap": true,
        "roundCap": true,
        "show": true,
      },
      "show": true,
      "startAngle": 210,
      "type": "gauge",
    },
  ],
  "title": Object {
    "text": "title-test",
  },
}
`;
